oauth2/src/main/java/com/linecorp/armeria/internal/common/auth/oauth2/AbstractOAuth2Request.java

```diff
@@ -0,0 +1,224 @@
+     * @throws TokenRequestException when the endpoint returns {code HTTP 400 (Bad Request)} status and the
+     *                               response payload contains the details of the error.
+     * @throws InvalidClientException when the endpoint returns {@code HTTP 401 (Unauthorized)} status, which
+     *                                typically indicates that client authentication failed (e.g.: unknown
+     *                                client, no client authentication included, or unsupported authentication
+     *                                method).
+     * @throws UnsupportedMediaTypeException if the media type of the response does not match the expected
+     *                                       (JSON).
+     */
+    @Nullable
+    protected T extractResults(AggregatedHttpResponse response, QueryParams requestFormData) {
+        final HttpStatus status = response.status();
+        switch (status.code()) {
+            case 200: // OK
+                // expected Content-Type: application/json;charset=UTF-8
+                validateContentType(response, MediaType.JSON);
+                return extractOkResults(response, requestFormData);
+            case 400: // Bad Request
+                // expected Content-Type: application/json;charset=UTF-8
+                validateContentType(response, MediaType.JSON);
+                throw onBadRequestError(response);
+            case 401: // Unauthorized
+                throw onUnauthorizedError(response);
+        }
+        throw new UnsupportedResponseException(status, response.contentUtf8());
+    }
+
+    /**
+     * Composes {@link TokenRequestException} upon 400 Bad Request response
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-5.2">[RFC6749], Section 5.2</a>.
+     * @param errorResponse response received from the server
+     * @return an instance of {@link TokenRequestException}
+     */
+    protected TokenRequestException onBadRequestError(AggregatedHttpResponse errorResponse) {
+        return TokenRequestException.parse(errorResponse.contentUtf8());
+    }
+
+    /**
+     * Composes {@link InvalidClientException} upon 401 Unauthorized response
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-5.2">[RFC6749], Section 5.2</a> (invalid_client).
+     * @param errorResponse response received from the server
+     * @return an instance of {@link InvalidClientException}
+     */
+    protected TokenRequestException onUnauthorizedError(AggregatedHttpResponse errorResponse) {
+        final StringBuilder messageBuilder = new StringBuilder().append(errorResponse.status());
+        final String wwwAuthenticate = errorResponse.headers().get(HttpHeaderNames.WWW_AUTHENTICATE);
+        if (wwwAuthenticate != null) {
+            messageBuilder.append(": ").append(wwwAuthenticate);
+        }
+        final HttpData errorResponseContents = errorResponse.content();
+        if (!errorResponse.content().isEmpty()) {
```

Comment:
Use the local variable?
```suggestion
        if (!errorResponseContents.isEmpty()) {
```

Reply:
Yes, this is how it's supposed to be. Fixed.